### PR TITLE
Add Kernel head

### PR DESCRIPTION
**Pull request by FANG-MING**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add Kernel head which is a rpn head used in K-net.

## Modification

Add Kernel head

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/detectors/knet.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn.functional as F
+
+from ..builder import DETECTORS
+from .two_stage import TwoStageDetector
+
+
+def sem2ins_masks(gt_sem_seg, num_thing_classes=80):
```
Comment (ZwwWayne) [inline mmdet/models/detectors/knet.py:9]: We can put this utils to mmdet.core.masks

File: mmdet/models/detectors/knet.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            size=[0, gt_sem_seg.shape[-2], gt_sem_seg.shape[-1]])
+    return labels.long(), masks.float()
+
+
+@DETECTORS.register_module()
+class KNet(TwoStageDetector):
+
+    def __init__(self,
+                 *args,
+                 num_thing_classes=80,
```
Comment (ZwwWayne) [inline mmdet/models/detectors/knet.py:54]: docstring

File: mmdet/models/detectors/knet.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        """
+        Args:
+            img (Tensor): of shape (N, C, H, W) encoding input images.
+                Typically these should be mean centered and std scaled.
+            img_metas (list[Dict]): list of image info dict where each dict
+                has: 'img_shape', 'scale_factor', 'flip', and may also contain
+                'filename', 'ori_shape', 'pad_shape', and 'img_norm_cfg'.
+                For details on the values of these keys see
+                `mmdet/datasets/pipelines/formatting.py:Collect`.
+            gt_bboxes (list[Tensor]): Ground truth bboxes for each image with
```
Comment (ZwwWayne) [inline mmdet/models/detectors/knet.py:85]: The order of arguments is inconsistent with that in the function

File: tests/test_utils/test_assigner.py [hunk 4]
```diff
@@ -628,3 +628,22 @@ def test_mask_hungarian_match_assigner():
         dice_cost=dict(type='DiceCost', weight=0.0, pred_act=True, eps=1.0))
     with pytest.raises(AssertionError):
         self = MaskHungarianAssigner(**assigner_cfg)
+
+    # test with solo_style mode of DiceLossCost which is not supported yet
+    assigner_cfg = dict(
```
Comment (ZwwWayne) [inline tests/test_utils/test_assigner.py:633]: The unit tests do not really test the module to be added in this PR.

Comment (FANG-MING) [reply tests/test_utils/test_assigner.py:633]: yes, it was modified, I will fix it.

> The unit tests do not really test the module to be added in this PR.
yes, it was modified, I will fix it.

File: mmdet/models/dense_heads/kernel_head.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+    def forward_train(self,
+                      feats,
+                      img_metas,
+                      gt_masks,
+                      gt_labels,
+                      gt_sem_seg=None,
+                      gt_sem_cls=None):
+        """
+        Args:
+            x (List[Tensor]): each of shape (N, C, H_i, W_i) \
```
Comment (chhluo) [inline mmdet/models/dense_heads/kernel_head.py:287]: no `\` for Args.

**Comment by ZwwWayne** [comment]
Please clean unnecessary files and resolve the conflict.
